syntax = "proto3";

package booksdb;

service BooksDatabase {
  rpc Read(ReadRequest) returns (ReadResponse);
  rpc Write(WriteRequest) returns (WriteResponse);
  rpc IncrementStock(StockUpdateRequest) returns (StockUpdateResponse);
  rpc DecrementStock(StockUpdateRequest) returns (StockUpdateResponse);
}

message ReadRequest {
  string key = 1;
}

message ReadResponse {
  string value = 1;
}

message WriteRequest {
  string key = 1;
  string value = 2;
}

message WriteResponse {
  bool success = 1;
}

message StockUpdateRequest {
  string key = 1;
  int32 amount = 2;
}

message StockUpdateResponse {
  bool success = 1;
}